### PR TITLE
Add 2D wave equation FNO training recipe

### PR DESCRIPTION
**Pull request by gpartin**
## Summary

This PR adds a complete Fourier Neural Operator training example for the **2D wave equation**, filling the hyperbolic PDE gap in PhysicsNeMo's example suite.

## Motivation

Current PhysicsNeMo examples focus on **elliptic** (Darcy flow) and **parabolic** (Navier-Stokes, heat) PDEs. The wave equation is a fundamental **hyperbolic** PDE critical for acoustics, seismology, and electromagnetic applications. This example provides a ready-to-use training recipe following the same patterns as \darcy_fno\.

## What's Included

| File | Purpose |
|------|---------|
| \	rain_fno_wave.py\ | Training script (Hydra config, distributed, checkpointing) |
| \wave_data.py\ | On-the-fly data generator using leapfrog finite differences |
| \alidator.py\ | Side-by-side visualization (initial/truth/prediction/error) |
| \config.yaml\ | 128x128 grid, 4-layer FNO, standard hyperparameters |
| \README.md\ | Problem description and usage instructions |
| \
equirements.txt\ | Minimal dependencies |

## Problem Setup

- **PDE**: \$\partial^2 u / \partial t^2 = c^2 \nabla^2 u\$
- **Domain**: \$[0,1]^2\$ with periodic boundaries
- **Input**: Random Fourier initial wavefield \(x,y,0)\$
- **Output**: Solution at target time \(x,y,T)\$
- **Solver**: Leapfrog finite-difference with Taylor expansion first step

## Design Decisions

- **On-the-fly generation**: Follows the Darcy2D pattern — unlimited unique training samples
- **No external dependencies**: Pure NumPy/PyTorch data generation, no Warp requirement
- **Standard FNO architecture**: Same \physicsnemo.models.fno.FNO\ as other examples
- **Configurable physics**: Wave speed, target time, and CFL number via Hydra config

## Usage

\\\ash
cd examples/wave/wave_fno
pip install -r requirements.txt
python train_fno_wave.py
\\\

## Testing

Verified data generator produces correct wave evolution (energy-conserving leapfrog with CFL-stable parameters).

### CODE REVIEW

**Comment by gpartin** [comment]
Thanks for the thorough review! I've pushed a commit addressing all feedback:

**Code quality fixes:**
- Removed unused `c2_ratio` variable (dead code)
- Fixed bare f-string `f"./checkpoints"` to plain string literal
- Reused existing `loss_fun` in `WaveValidator` instead of creating a duplicate `MSELoss`

**Robustness:**
- Changed `device` default from `"cuda"` to `"cpu"` in both `generate_wave_batch()` and `WaveDataLoader` so the example runs without CUDA. Type widened to `str | torch.device`.
- Added input validation for `resolution`, `batch_size`, `wave_speed`, `cfl`, and `target_time` (fail-fast with clear errors)

**Reproducibility:**
- Added `seed` parameter to `WaveDataLoader`; each batch gets `base_seed + batch_counter` so runs are deterministic when a seed is provided

**Vectorization note:**
- Added a comment explaining the per-sample loop trade-off (simplicity/low memory vs. throughput). A fully vectorized or GPU-based solver is noted as the path for higher throughput.

Re the leapfrog step count observation — the current logic is correct: 1 Taylor bootstrap step + (n_steps-1) leapfrog steps = n_steps total advances, landing at t = T exactly.

Let me know if anything else needs attention!
